### PR TITLE
windows support

### PR DESCRIPTION
**Pull request by mattn**
ref #1

### CODE REVIEW

**Comment by c-bata** [comment]
Hi, @mattn

Thanks for supporting windows! I want to try this changes as soon as possible, but unfortunatelyl I don't have windows environment.
Please wait to merge it... I'm gonna prepare a windows machine.

Thanks.

**Comment by mattn** [comment]
Current implementation of Windows port using go-tty. This provide Size(), ReadRune(). So I guess some of code for UN*X can be replaced using go-tty. (if you want)

File: prompt_windows.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+	log.Printf("[INFO] readBuffer start")
+	for {
+		time.Sleep(10 * time.Millisecond)
+		select {
+		case <-stopCh:
+			log.Print("[INFO] stop readBuffer")
+			return
+		default:
+			if r, err := p.in.(*VT100Parser).tty.ReadRune(); err == nil {
```
Comment (c-bata) [inline prompt_windows.go]: Hmm... I need to refactor or need to remove `ConsoleParser` interface. I will do it after merging this PR.

File: vt100_output_windows.go [hunk 2]
```diff
@@ -0,0 +1,299 @@
+// +build windows
```
Comment (c-bata) [inline vt100_output_windows.go]: I thought about windows console doesn't understand vt100 escape sequences. but It's not true?
If it is true, I think this file should be named `windows_output.go` and should rename `VT100Writer`.

I'll check this later.

File: signal_windows.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+func handleSignals(in ConsoleParser, exitCh chan int, winSizeCh chan *WinSize) {
```
Comment (mattn) [inline signal_windows.go]: On Windows, ReadConsoleInput and WINDOW_BUFFER_SIZE_EVENT can be used for this.

File: prompt_windows.go (hunk 1 above)
Comment (mattn) [inline prompt_windows.go]: Yes right, it prefer to add reader.

File: vt100_output_windows.go (hunk 2 above)
Comment (mattn) [inline vt100_output_windows.go]: go-colorable handles most of escape sequence(s) on Windows.

https://github.com/mattn/go-colorable/blob/master/colorable_windows.go

File: signal_windows.go (hunk 3 above)
Comment (c-bata) [inline signal_windows.go]: I see. thanks!

File: vt100_output_windows.go (hunk 2 above)
Comment (c-bata) [inline vt100_output_windows.go]: Wow, it's great library to support windows.
It seems to reduce the duplicate of code.

File: prompt_windows.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	"time"
+	"unicode/utf8"
+)
+
+func (p *Prompt) readBuffer(bufCh chan []byte, stopCh chan struct{}) {
+	buf := make([]byte, 1024)
+
+	log.Printf("[INFO] readBuffer start")
+	for {
+		time.Sleep(10 * time.Millisecond)
```
Comment (hibooboo2) [inline prompt_windows.go]: Why is the sleep needed here? Is this something windows specific?

Comment (hibooboo2) [reply prompt_windows.go]: Nvm I see its in posix as well. So not windows specific. Then y the sleep. Jw.

File: signal_windows.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+		case syscall.SIGTERM: // kill -SIGTERM XXXX
+			log.Println("[SIGNAL] Catch SIGTERM")
+			exitCh <- 1
+
+		case syscall.SIGQUIT: // kill -SIGQUIT XXXX
+			log.Println("[SIGNAL] Catch SIGQUIT")
+			exitCh <- 0
+
+		default:
+			time.Sleep(10 * time.Millisecond)
```
Comment (hibooboo2) [inline signal_windows.go]: Ok, not trying to Bombard maybe I am still missing gap for go best practices. Why the sleep here?

**Comment by hibooboo2** [comment]
@mattn So looking at the diff between vt100_output.go and vt100_output_windows.go there is very little different. Why not have 3 files vt100_output.go vt_100_output_windows.go and vt_100_output_notwindows.go or vt_100_output_posix.go and have the code that is the same in output.go and the differences in the posix and windows files? You can even have the struct defs in the different files and it still works properly. Just a thought. Would reduce size and make it easier to maintain moving forward.

**Comment by hibooboo2** [comment]
Really would love this pr to get merged though. Tried to use this on windows for a project yesterday and had some wierd things w/ out this pr.

**Comment by hibooboo2** [comment]
Also I tried using this pr as is on my windows machine to compile and it will not compile.
```
$ GOOS=windows go build
# github.com/c-bata/go-prompt
../../c-bata/go-prompt/emacs.go:117: unknown field 'fd' in struct literal of type VT100Writer
```

**Comment by hibooboo2** [comment]
This change fixes it: 
```
$ git diff HEAD
diff --git a/emacs.go b/emacs.go
index 87138a3..458d060 100644
--- a/emacs.go
+++ b/emacs.go
@@ -1,9 +1,5 @@
 package prompt

-import (
-       "syscall"
-)
-
 /*

 ========
@@ -114,7 +110,7 @@ var emacsKeyBindings = []KeyBind{
        {
                Key: ControlL,
                Fn: func(buf *Buffer) {
-                       out := &VT100Writer{fd: syscall.Stdout}
+                       out := NewVT100StandardOutputWriter()
                        out.EraseScreen()
                        out.CursorGoTo(0, 0)
                },
```

File: prompt_windows.go (hunk 4 above)
Comment (mattn) [reply prompt_windows.go]: I kept it original code.

File: signal_windows.go (hunk 5 above)
Comment (mattn) [reply signal_windows.go]: ditto

**Comment by mattstratton** [comment]
I am very interested in seeing this PR get merged (if it works) - I came across `go-prompt` after hearing about it on the Go Time podcast, and would love to use it in a cli I'm building...sadly it has to include Windows support!

**Comment by hibooboo2** [comment]
This pr is still a bit buggy on windows.. Using the plain cmd   prompt...

**Pull request closed without merging**